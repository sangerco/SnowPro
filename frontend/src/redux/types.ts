export interface UserData {
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
  }


export const FETCH_USER_DATA_REQUEST = 'FETCH_USER_DATA_REQUEST';
export const FETCH_USER_DATA_SUCCESS = 'FETCH_USER_DATA_SUCCESS';
export const FETCH_USER_DATA_FAILURE = 'FETCH_USER_DATA_FAILURE';


export interface FetchUserDataRequestAction {
    type: typeof FETCH_USER_DATA_REQUEST
}

export interface FetchUserDataSuccessAction {
    type: typeof FETCH_USER_DATA_SUCCESS;
    payload: UserData;
}

export interface FetchUserDataFailureAction {
    type: typeof FETCH_USER_DATA_FAILURE;
    payload: string;
}

export type UserActionTypes = 
    FetchUserDataRequestAction
    | FetchUserDataSuccessAction
    | FetchUserDataFailureAction;