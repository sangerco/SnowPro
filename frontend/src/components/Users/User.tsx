import React, { useEffect } from 'react';
import { Dimmer, Loader } from "semantic-ui-react";
import { connect } from 'react-redux';
import { fetchUserData } from '../../redux/actions/userActions';
import ShowPhotos from '../Media/ShowPhotos';
import ShowVideos from '../Media/ShowVideos';
import { Container, Header, Grid, Image } from 'semantic-ui-react';
import FavMountain from '../SkiAreas/FavMountain';
import { useParams } from 'react-router';

interface UserData {
    id: string;
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
}

interface UserProps {
    user: UserData | null;
    loading: boolean;
    error: string | null;
    fetchUserData: ( username: string) => void;
}

const User: React.FC<UserProps> = ({ user, loading, error, fetchUserData }) => {
    const { username } = useParams();

    useEffect(() => {
        fetchUserData(username as string);
    }, [username, fetchUserData]);

    if(loading) {
        return (
            <div>
                <Dimmer active>
                    <Loader>Loading...</Loader>
                </Dimmer>
            </div>
        )
    }

    if(error) {
        return (
            <div>Error: {error}</div>
        )
    }

    if(user) {
        return (
            <Container>
                <Grid relaxed>
                    <Grid.Row>
                        <Grid.Column width={4}>
                            <Image src={user.avatar} size='small' />
                        </Grid.Column>
                        <Grid.Column width={8}>
                            <Header as='h2' textAlign='right'>{user.username}</Header>
                            <Header.Subheader>
                                {user.firstName && user.firstName !== '' ? user.firstName : ''}
                                {user.lastName && user.lastName !== '' ? user.lastName : ''}
                            </Header.Subheader>
                        </Grid.Column>
                        <Grid.Column width={4}></Grid.Column>
                    </Grid.Row>
                    <Grid.Row>
                        <Grid.Column width={4}></Grid.Column>
                        <Grid.Column width={8}><p>{user.bio}</p>
                        </Grid.Column>
                        <Grid.Column width={4}>
                            {user.favMountains.length > 0 ? 
                                <FavMountain userId={user.id} /> :
                                <p>This user has no favorited mountains yet!</p>}
                        </Grid.Column>
                    </Grid.Row>
                    <Grid.Row>
                        <Grid.Column width={2}></Grid.Column>
                        <Grid.Column width={12}><ShowPhotos username={username as string} /></Grid.Column>
                        <Grid.Column width={2}></Grid.Column>
                    </Grid.Row>
                    <Grid.Row>
                        <Grid.Column width={2}></Grid.Column>
                        <Grid.Column width={12}><ShowVideos username={username as string} /></Grid.Column>
                        <Grid.Column width={2}></Grid.Column>
                    </Grid.Row>
                </Grid>

            </Container>


             //     {user.favMountains.length > 0 ? user.favMountains.map(fm => <p key={fm}>{fm}</p>) : <p>No favorite mountains yet.</p>}
            // </div>
        )
    }

    return null;
}

const mapStateToProps = (state: any) => ({
    user: state.user.data,
    loading: state.user.loading,
    error: state.user.error
})

const mapDispatchToProps = {
    fetchUserData
}

export default connect(mapStateToProps, mapDispatchToProps)(User);