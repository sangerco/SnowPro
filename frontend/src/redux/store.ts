import { Action, compose } from "redux";
import { ThunkAction, configureStore } from "@reduxjs/toolkit";
import { logger } from "redux-logger";
import rootReducer from "./reducers/rootReducer";

const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(logger, composeEnhancers),
});
