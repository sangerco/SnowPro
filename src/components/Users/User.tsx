import React from "react";

interface UserProps {
  user: {
    id: string;
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    bio?: string;
    avatar?: string;
    isAdmin: boolean;
  };
}

const User: React.FC<UserProps> = ({ user }) => {
  return (
    <li key={user.id}>
      <p>{user.id}</p>
      <p>{user.username}</p>
      <p>{user.firstName}</p>
      <p>{user.lastName}</p>
      <p>{user.email}</p>
      <p>{user.bio}</p>
      <p>{user.avatar}</p>
      <p>{user.isAdmin}</p>
    </li>
  );
};

export default User;
