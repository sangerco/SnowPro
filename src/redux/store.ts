import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
