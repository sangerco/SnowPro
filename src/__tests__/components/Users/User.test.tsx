/* eslint-disable testing-library/no-node-access */
import React from "react";
import User from "../../../components/Users/User";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import { BrowserRouter } from "react-router-dom";

describe("test User", () => {
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <User
            user={{
              id: "11",
              username: "testuser",
              firstName: "Test",
              lastName: "User",
              email: "test@user.com",
              bio: "test user bio",
              avatar: "https://testdata.com/testavatar.jpg",
              isAdmin: false,
            }}
          />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <User
            user={{
              id: "11",
              username: "testuser",
              firstName: "Test",
              lastName: "User",
              email: "test@user.com",
              bio: "test user bio",
              avatar: "https://testdata.com/testavatar.jpg",
              isAdmin: false,
            }}
          />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
