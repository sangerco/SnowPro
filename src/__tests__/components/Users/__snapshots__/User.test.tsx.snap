// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test User should match snapshot 1`] = `
<DocumentFragment>
  <li>
    <p>
      11
    </p>
    <p>
      testuser
    </p>
    <p>
      Test
    </p>
    <p>
      User
    </p>
    <p>
      test@user.com
    </p>
    <p>
      test user bio
    </p>
    <p>
      https://testdata.com/testavatar.jpg
    </p>
    <p />
  </li>
</DocumentFragment>
`;
